src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+                tail = firstNode;
+                head = firstNode;
+                size++;
+                return;
+            }
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        }
+        if (index > 0 && index < size) {
+            Node<T> current = searchNodeByIndex(index);
+            Node<T> newNode = new Node<>(current.prev, value, current);
+            current.prev.next = newNode;
+            current.prev = newNode;
+            size++;
+        }
+        if (index == size) {
+            add(value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = searchNodeByIndex(index);
+        return current.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T replacedElement;
+        Node<T> current = searchNodeByIndex(index);
+        replacedElement = current.item;
+        current.item = value;
+        return replacedElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = searchNodeByIndex(index);
+        if (size == 1 && index == 0) {
+            size--;
+        } else if (index == 0) {
+            current.next.prev = null;
+            head = current.next;
+            size--;
+        } else if (index > 0 && index < size - 1) {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+            size--;
+        } else if (index == size - 1) {
+            current.prev.next = null;
+            tail = current.prev;
+            size--;
+        }
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.item == object || current.item.equals(object)) {
+                remove(i);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void checkIndexForAddMethod(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("IndexOutOfBoundsException");
+        }
+    }
+
+    public void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("IndexOutOfBoundsException");
+        }
+    }
+
+    public Node<T> searchNodeByIndex(int index) {
```

Comment:
checklist: Don’t iterate your nodes from the head to tail or vice versa.
Pay attention to the index you receive as an input, you can iterate only first or second half of the list, depending on the index value. That will boost your lists’ performance.